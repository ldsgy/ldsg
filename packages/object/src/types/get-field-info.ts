import { PlatformParams } from "@ldsg/field-type";
import { GeneralResourceSettings } from "@ldsg/types";

export interface FieldInfo extends GeneralResourceSettings {
  /**
   * Field Name
   */
  name: string;

  /**
   * Type Info
   */
  typeInfo: any;
}

interface GetFieldInfoRes {
  fieldInfo: FieldInfo;
}

export type GetFieldInfo = (params: PlatformParams) => GetFieldInfoRes;
