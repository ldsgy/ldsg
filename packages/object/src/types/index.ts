export * from "./get-field-info";
export * from "./specific-resource-settings";
