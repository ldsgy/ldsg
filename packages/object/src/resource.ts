import { PlatformsParams } from "@ldsg/field-type";
import { Resource } from "@ldsg/resource";
import { GeneralResourceSettings } from "@ldsg/types";
import _ from "lodash";
import {
  FieldInfo,
  GetFieldInfo,
  ObjectSpecificResourceSettings,
} from "./types";

export interface ExtendedObjectInfo {
  /**
   * Field Info List
   */
  fieldInfoList: FieldInfo[];
}

export interface GetExtendedObjectInfoRes {
  /**
   * Extended Object Info
   */
  extendedObjectInfo: ExtendedObjectInfo;
}

type GetExtendedObjectInfo = (
  params?: PlatformsParams
) => GetExtendedObjectInfoRes;

export interface ObjectInfo extends GeneralResourceSettings {
  /**
   * Object Name
   */
  name: string;

  /**
   * Field Info List
   */
  fieldInfoList: FieldInfo[];
}

export interface GetObjectInfoRes {
  /**
   * Object Info
   */
  objectInfo: ObjectInfo;
}

type GetObjectInfo = (params?: PlatformsParams) => GetObjectInfoRes;

export class ObjectResource extends Resource<ObjectSpecificResourceSettings> {
  getExtendedObjectInfo: GetExtendedObjectInfo = (params) => {
    const {
      settings: { extendedResourceIds },
      getFilteredResource,
    } = this;

    const fieldInfoListGroup = _.map(
      extendedResourceIds,
      (extendedResourceId) => {
        const { resource } = getFilteredResource<ObjectResource>({
          id: extendedResourceId,
        });

        const getObjectInfoRes = resource?.getObjectInfo(params);

        const res = getObjectInfoRes?.objectInfo.fieldInfoList || [];

        return res;
      }
    );

    const fieldInfoList = _.flatten(fieldInfoListGroup);

    const extendedObjectInfo: ExtendedObjectInfo = {
      fieldInfoList,
    };

    const res = {
      extendedObjectInfo,
    };

    return res;
  };

  getObjectInfo: GetObjectInfo = (params) => {
    const {
      id,
      settings: { title, description, name },
      getExtendedObjectInfo,
      getFilteredResources,
    } = this;

    const { extendedObjectInfo } = getExtendedObjectInfo(params);

    const { fieldInfoList: extendedFieldInfoList } = extendedObjectInfo;

    const { resources } = getFilteredResources<{ getFieldInfo?: GetFieldInfo }>(
      {
        parentId: id,
      }
    );

    const mapRes = resources.map((resource) => {
      const fieldInfo = resource.getFieldInfo?.(params);

      return fieldInfo;
    });

    const filterRes = _.filter(mapRes, (value) => !_.isUndefined(value));

    const currentFieldInfoList = _.map(filterRes, "fieldInfo");

    const fieldInfoList = extendedFieldInfoList.concat(currentFieldInfoList);

    const objectInfo: ObjectInfo = {
      title,
      description,
      name,
      fieldInfoList,
    };

    const res = {
      objectInfo,
    };

    return res;
  };
}
