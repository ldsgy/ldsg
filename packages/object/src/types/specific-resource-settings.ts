import { SpecificResourceSettings } from "@ldsg/types";

export interface ObjectSpecificResourceSettings
  extends SpecificResourceSettings {}
