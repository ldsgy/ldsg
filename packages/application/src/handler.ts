import { Handler } from "@ldsg/handler";
import { Manifest } from "@ldsg/resource";
import { ApplicationResource } from "./resource";
import { ApplicationResourceSettings } from "./types";

export const handler: Handler<
  [Manifest.Resource<ApplicationResourceSettings>]
> = (params) => {
  return new ApplicationResource(params);
};
