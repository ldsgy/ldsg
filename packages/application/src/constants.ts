import { ROOT_RESOURCE_KIND } from "@ldsg/constants";
import { HandlerSpecificResourceSettings } from "@ldsg/handler";
import { ResourceDefinitionSpecificResourceSettings } from "@ldsg/resource-definition";
import { GeneralResourceSettings } from "@ldsg/types";

/**
 * Instantiate Resource Handler General Resource Settings
 * 实例化资源处理程序普通资源配置
 */
export const INSTANTIATE_RESOURCE_HANDLER_GENERAL_RESOURCE_SETTINGS: GeneralResourceSettings =
  {
    title: "应用类型资源处理程序",
    description: "",
  };

/**
 * Instantiate Resource Handler Resource Settings
 * 实例化资源处理程序特殊资源配置
 * 框架通过此实例化此资源
 */
export const INSTANTIATE_RESOURCE_HANDLER_SPECIFIC_RESOURCE_SETTINGS: HandlerSpecificResourceSettings =
  {
    code: `export * from "@ldsg/application";`,
    dependencies: [
      {
        name: "@ldsg/application",
      },
    ],
  };

export const RESOURCE_DEFINITION_GENERAL_RESOURCE_SETTINGS: GeneralResourceSettings =
  { title: "应用类型资源定义", description: "" };

export const RESOURCE_DEFINITION_SPECIFIC_RESOURCE_SETTINGS: ResourceDefinitionSpecificResourceSettings =
  {
    kind: "application",
    parentKind: ROOT_RESOURCE_KIND,
  };
