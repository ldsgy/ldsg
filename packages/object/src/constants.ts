import { HandlerSpecificResourceSettings } from "@ldsg/handler";
import { ResourceDefinitionSpecificResourceSettings } from "@ldsg/resource-definition";
import { GeneralResourceSettings } from "@ldsg/types";

export const HANDLER_GENERAL_RESOURCE_SETTINGS: GeneralResourceSettings = {
  title: "对象类型资源处理程序",
  description: "",
};

/**
 * Handler Resource Settings
 * 此类型资源的子级处理程序资源的配置
 * 如无或为 undefined，则说明无需子级处理程序。
 */
export const HANDLER_SPECIFIC_RESOURCE_SETTINGS: HandlerSpecificResourceSettings =
  {
    code: `export * from "@ldsg/object";`,
    dependencies: [
      {
        name: "@ldsg/object",
      },
    ],
  };

export const RESOURCE_DEFINITION_GENERAL_RESOURCE_SETTINGS: GeneralResourceSettings =
  { title: "对象类型资源定义", description: "" };

export const RESOURCE_DEFINITION_SPECIFIC_RESOURCE_SETTINGS: ResourceDefinitionSpecificResourceSettings =
  {
    kind: "object",
    parentKind: "",
  };
