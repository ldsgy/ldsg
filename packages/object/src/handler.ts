import { Handler, HandlerExtendedResourceSettings } from "@ldsg/handler";
import {
  InstantiateResourceParams,
  InstantiateResourceRes,
} from "@ldsg/resource";
import { instantiateResource } from "./instantiate-resource";
import { ObjectResource } from "./resource";
import { ObjectSpecificResourceSettings } from "./types";

export const handler: Handler<
  [
    InstantiateResourceParams<
      HandlerExtendedResourceSettings<ObjectSpecificResourceSettings>
    >
  ],
  InstantiateResourceRes<ObjectResource>
> = instantiateResource;
