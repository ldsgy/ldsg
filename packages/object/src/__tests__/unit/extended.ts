import { ROOT_RESOURCE_ID } from "@ldsg/constants";
import {
  FieldTypeResource,
  stringFieldTypeHandlerResourceRecord,
  stringFieldTypeResourceRecord,
} from "@ldsg/field-type";
import { HandlerResource } from "@ldsg/handler";
import { Resource } from "@ldsg/resource";
import { SpecificResourceSettings } from "@ldsg/types";
import { ObjectResource } from "../../resource";
import { FieldInfo, GetFieldInfo } from "../../types";

interface ASpecificResourceSettings extends SpecificResourceSettings {
  /**
   * Field Name
   */
  name: string;

  /**
   * Field Type Resource ID
   */
  fieldTypeResourceId: string;

  /**
   * Field Properties
   */
  properties: any;
}

class MockObjectFieldResource extends Resource<ASpecificResourceSettings> {
  getFieldInfo: GetFieldInfo = (params) => {
    const {
      id,
      settings: { title, description, name, properties },
      getResourcesFromSettings,
    } = this;

    const { fieldTypeResource } = getResourcesFromSettings();

    const { fieldTypeInfo } = (
      fieldTypeResource as FieldTypeResource
    ).getFieldTypeInfo({
      ...params,
      fieldProperties: properties,
    });

    const fieldInfo: FieldInfo = {
      id,
      title,
      description,
      name,
      typeInfo: fieldTypeInfo,
    };

    const res = {
      fieldInfo,
    };

    return res;
  };
}

test("object", () => {
  new HandlerResource(stringFieldTypeHandlerResourceRecord);

  new FieldTypeResource(stringFieldTypeResourceRecord);

  const test1Object = new ObjectResource({
    id: "test-1-object",
    kind: "object",
    parentId: ROOT_RESOURCE_ID,
    settings: {
      title: "测试1对象",
      description: "",
      name: "test-1",
    },
  });

  const test2Object = new ObjectResource({
    id: "test-2-object",
    kind: "object",
    parentId: ROOT_RESOURCE_ID,
    settings: {
      title: "测试2对象",
      description: "",
      name: "test-2",
      extendedResourceIds: ["test-1-object"],
    },
  });

  /**
   * 测试 1 对象的字段
   */
  {
    new MockObjectFieldResource({
      id: "test-1-object-test-1-object-field",
      kind: "object_field",
      parentId: "test-1-object",
      settings: {
        title: "测试1对象测试1对象字段",
        description: "",
        name: "test-1",
        fieldTypeResourceId: "test-field-type",
        properties: {
          max: "10",
        },
      },
    });

    new MockObjectFieldResource({
      id: "test-1-object-test-2-object-field",
      kind: "object_field",
      parentId: "test-1-object",
      settings: {
        title: "测试1对象测试2对象字段",
        description: "",
        name: "test2",
        fieldTypeResourceId: "test-field-type",
        properties: {
          max: "10",
        },
      },
    });

    new MockObjectFieldResource({
      id: "test-1-object-test-3-object-field",
      kind: "object_field",
      parentId: "test-1-object",
      settings: {
        title: "测试1对象测试3对象字段",
        description: "",
        name: "test3",
        fieldTypeResourceId: "test-field-type",
        properties: {
          max: "10",
        },
      },
    });
  }

  /**
   * 测试 2 对象的字段
   */
  {
    new MockObjectFieldResource({
      id: "test-2-object-test-1-object-field",
      kind: "object_field",
      parentId: "test-2-object",
      settings: {
        title: "测试2对象测试1对象字段",
        description: "",
        name: "test-1",
        fieldTypeResourceId: "test-field-type",
        properties: {
          max: "10",
        },
      },
    });

    new MockObjectFieldResource({
      id: "test-2-object-test-4-object-field",
      kind: "object_field",
      parentId: "test-2-object",
      settings: {
        title: "测试2对象测试4对象字段",
        description: "",
        name: "test-4",
        fieldTypeResourceId: "test-field-type",
        properties: {
          max: "10",
        },
      },
    });

    new MockObjectFieldResource({
      id: "test-2-object-test-5-object-field",
      kind: "object_field",
      parentId: "test-2-object",
      settings: {
        title: "测试2对象测试5对象字段",
        description: "",
        name: "test-5",
        fieldTypeResourceId: "test-field-type",
        properties: {
          max: "10",
        },
      },
    });
  }

  const getObjectInfoRes1 = test1Object.getObjectInfo();

  expect(getObjectInfoRes1).toMatchSnapshot();

  const getObjectInfoRes2 = test2Object.getObjectInfo();

  expect(getObjectInfoRes2).toMatchSnapshot();
});
