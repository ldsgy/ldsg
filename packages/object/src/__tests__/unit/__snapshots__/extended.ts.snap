// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`object 1`] = `
{
  "objectInfo": {
    "description": "",
    "fieldInfoList": [
      {
        "description": "",
        "id": "test-1-object-test-1-object-field",
        "name": "test-1",
        "title": "测试1对象测试1对象字段",
        "typeInfo": {
          "platformToTypeMap": {
            "COMMON": {
              "max": "10",
              "type": "string",
            },
            "GRAPHQL": undefined,
            "JSON": undefined,
            "MONGOOSE": undefined,
          },
        },
      },
      {
        "description": "",
        "id": "test-1-object-test-2-object-field",
        "name": "test2",
        "title": "测试1对象测试2对象字段",
        "typeInfo": {
          "platformToTypeMap": {
            "COMMON": {
              "max": "10",
              "type": "string",
            },
            "GRAPHQL": undefined,
            "JSON": undefined,
            "MONGOOSE": undefined,
          },
        },
      },
      {
        "description": "",
        "id": "test-1-object-test-3-object-field",
        "name": "test3",
        "title": "测试1对象测试3对象字段",
        "typeInfo": {
          "platformToTypeMap": {
            "COMMON": {
              "max": "10",
              "type": "string",
            },
            "GRAPHQL": undefined,
            "JSON": undefined,
            "MONGOOSE": undefined,
          },
        },
      },
    ],
    "name": "test-1",
    "title": "测试1对象",
  },
}
`;

exports[`object 2`] = `
{
  "objectInfo": {
    "description": "",
    "fieldInfoList": [
      {
        "description": "",
        "id": "test-1-object-test-1-object-field",
        "name": "test-1",
        "title": "测试1对象测试1对象字段",
        "typeInfo": {
          "platformToTypeMap": {
            "COMMON": {
              "max": "10",
              "type": "string",
            },
            "GRAPHQL": undefined,
            "JSON": undefined,
            "MONGOOSE": undefined,
          },
        },
      },
      {
        "description": "",
        "id": "test-1-object-test-2-object-field",
        "name": "test2",
        "title": "测试1对象测试2对象字段",
        "typeInfo": {
          "platformToTypeMap": {
            "COMMON": {
              "max": "10",
              "type": "string",
            },
            "GRAPHQL": undefined,
            "JSON": undefined,
            "MONGOOSE": undefined,
          },
        },
      },
      {
        "description": "",
        "id": "test-1-object-test-3-object-field",
        "name": "test3",
        "title": "测试1对象测试3对象字段",
        "typeInfo": {
          "platformToTypeMap": {
            "COMMON": {
              "max": "10",
              "type": "string",
            },
            "GRAPHQL": undefined,
            "JSON": undefined,
            "MONGOOSE": undefined,
          },
        },
      },
      {
        "description": "",
        "id": "test-2-object-test-1-object-field",
        "name": "test-1",
        "title": "测试2对象测试1对象字段",
        "typeInfo": {
          "platformToTypeMap": {
            "COMMON": {
              "max": "10",
              "type": "string",
            },
            "GRAPHQL": undefined,
            "JSON": undefined,
            "MONGOOSE": undefined,
          },
        },
      },
      {
        "description": "",
        "id": "test-2-object-test-4-object-field",
        "name": "test-4",
        "title": "测试2对象测试4对象字段",
        "typeInfo": {
          "platformToTypeMap": {
            "COMMON": {
              "max": "10",
              "type": "string",
            },
            "GRAPHQL": undefined,
            "JSON": undefined,
            "MONGOOSE": undefined,
          },
        },
      },
      {
        "description": "",
        "id": "test-2-object-test-5-object-field",
        "name": "test-5",
        "title": "测试2对象测试5对象字段",
        "typeInfo": {
          "platformToTypeMap": {
            "COMMON": {
              "max": "10",
              "type": "string",
            },
            "GRAPHQL": undefined,
            "JSON": undefined,
            "MONGOOSE": undefined,
          },
        },
      },
    ],
    "name": "test-2",
    "title": "测试2对象",
  },
}
`;
